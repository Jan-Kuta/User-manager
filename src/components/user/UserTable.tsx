import React, {useState} from 'react';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';
import PersonAddIcon from '@material-ui/icons/PersonAdd';
import UsersIcon from '@material-ui/icons/SupervisedUserCircle';
import Hidden from '@material-ui/core/Hidden';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TablePagination from '@material-ui/core/TablePagination';
import Paper from '@material-ui/core/Paper';
import {User} from "../../types/userTypes";
import {UserTableRow} from "./UserTableRow";
import {useTranslation} from "react-i18next";
import makeStyles from '@material-ui/core/styles/makeStyles';
import {Link} from "react-router-dom";
import {UserEditDialog} from "./UserEditDialog";
import {UserDeleteDialog} from "./UserDeleteDialog";

type Props = {
    users: User[]
}

const useStyles = makeStyles((theme) => ({
    toolbarTitle: {
        display: 'flex',
        flex: '1 1 100%'
    },
    toolbarTitleIcon: {
        marginRight: theme.spacing(2)
    },
    tableContainer: {
        maxHeight: '80vh'
    },
    tableHeader: {
        '& th': {
            backgroundColor: theme.palette.background.paper
        }
    }
}));

export const UserTable = (props: Props) => {
    const { users } = props;
    const { t } = useTranslation();
    const classes = useStyles(props);
    const [page, setPage] = useState(0);
    const [isOpenUserEditDialog, setOpenUserEditDialog] = useState(false);
    const [isOpenUserDeleteDialog, setOpenUserDeleteDialog] = useState(false);
    const [managingUser, setManagingUser] = useState<User | null>(null);
    const [rowsPerPage, setRowsPerPage] = useState(5);

    const handleChangePage = (event: unknown, newPage: number) => {
        setPage(newPage);
    };

    const handleChangeRowsPerPage = (event: React.ChangeEvent<HTMLInputElement>) => {
        setRowsPerPage(parseInt(event.target.value, 10));
        setPage(0);
    };

    const handleOpenUserEditDialog = (user: User) => {
        setManagingUser(user);
        setOpenUserEditDialog(true);
    };

    const handleOpenUserDeleteDialog = (user: User) => {
        setManagingUser(user);
        setOpenUserDeleteDialog(true);
    };

    const handleCloseUserEditDialog = () => setOpenUserEditDialog(false);

    const handleCloseUserDeleteDialog = () => setOpenUserDeleteDialog(false);

    return (
        <>
            <Paper>
                <Toolbar>
                        <div className={classes.toolbarTitle}>
                            <UsersIcon className={classes.toolbarTitleIcon} />
                            <Typography variant="h6" id="tableTitle" component="div">
                                {t('User list')}
                            </Typography>
                        </div>
                        <Link to='/users/create'>
                            <Tooltip placement="left" title={t('Add user').toString()}>
                                <IconButton aria-label="add user">
                                    <PersonAddIcon />
                                </IconButton>
                            </Tooltip>
                        </Link>
                </Toolbar>
                <TableContainer className={classes.tableContainer}>
                    <Table aria-label="collapsible table" size="small" stickyHeader>
                        <TableHead className={classes.tableHeader}>
                            <TableRow>
                                <TableCell />
                                <TableCell align="right">{t('Firstname')}</TableCell>
                                <TableCell align="right">{t('Lastname')}</TableCell>
                                <Hidden xsDown>
                                    <TableCell align="right">{t('Username')}</TableCell>
                                </Hidden>
                                <TableCell align="right">{t('State')}</TableCell>
                                <Hidden xsDown>
                                    <TableCell align="right">{t('Actions')}</TableCell>
                                </Hidden>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {users.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage).map((user) => (
                                <UserTableRow
                                    key={user.id}
                                    user={user}
                                    openUserEditDialog={handleOpenUserEditDialog}
                                    openUserDeleteDialog={handleOpenUserDeleteDialog}
                                />
                            ))}
                        </TableBody>
                    </Table>
                </TableContainer>
                <TablePagination
                    rowsPerPageOptions={[5, 10, 25]}
                    backIconButtonText={t('Back')}
                    nextIconButtonText={t('Next')}
                    labelRowsPerPage={t('Rows per page')}
                    labelDisplayedRows={({ from, to, count }) => `${from} - ${to} ${t('Of')} ${count}`}
                    component="div"
                    count={users.length}
                    rowsPerPage={rowsPerPage}
                    page={page}
                    onChangePage={handleChangePage}
                    onChangeRowsPerPage={handleChangeRowsPerPage}
                />
            </Paper>
            {managingUser && <UserEditDialog userToEdit={managingUser} open={isOpenUserEditDialog} onClose={handleCloseUserEditDialog}/>}
            {managingUser && <UserDeleteDialog userToEdit={managingUser} open={isOpenUserDeleteDialog} onClose={handleCloseUserDeleteDialog}/>}
        </>
    )
};
