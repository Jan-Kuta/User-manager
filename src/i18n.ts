import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// the translations
// (tip move them in a JSON file and import them)
const resources = {
    en: {
        translation: {
            "Language": "Language",
            "Note list": "Note list",
            "New note": "New note",
            "Note detail": "Note detail",
            "Title": "Title",
            "Delete": "Delete",
            "Edit": "Edit",
            "Dialog title": "Do you realy want to delete this item?",
            "Dialog description": "It can't be undone.",
            "Cancel": "Cancel",
            "OK": "OK",
            "Note edit": "Note edit",
            "Submit": "Submit",
            "User list": "User list",
            "User detail": "User detail",
            "Create user": "New user",
            "STATE_NEW": "New",
            "STATE_VALIDATED": "Validated",
            "STATE_VIP": "VIP",
            "Add user": "Add user",
            "Firstname": "Firstname",
            "Lastname": "Lastname",
            "Username": "Username",
            "State": "State",
            "Actions": "Actions"
        }
    },
    cs: {
        translation: {
            "Language": "Jazyk",
            "Note list": "Poznámky",
            "New note": "Nová poznámka",
            "Note detail": "Detail poznámky",
            "Title": "Popisek",
            "Delete": "Smazat",
            "Edit": "Upravit",
            "Dialog title": "Opravdu si přejete smazat tuto položku?",
            "Dialog description": "Tato akce nelze vrátit zpět.",
            "Cancel": "Zrušit",
            "OK": "Potvrdit",
            "Note edit": "Úprava poznámky",
            "Submit": "Uložit",
            "User list": "Seznam uživatelů",
            "User detail": "Další údaje",
            "Create user": "Nový uživatel",
            "Edit user": "Editace uživatele",
            "STATE_NEW": "Nový",
            "STATE_VALIDATED": "Ověřený",
            "STATE_VIP": "VIP",
            "Add user": "Přidat uživatele",
            "Basic info": "Základní údaje",
            "Firstname": "Jméno",
            "Lastname": "Příjmení",
            "Username": "Uživatelské jméno",
            "State": "Stav",
            "Actions": "Akce",
            "Email": "E-mail",
            "Phone": "Telefonní číslo",
            "Street": "Ulice",
            "City": "Město",
            "Zip": "Směrovací číslo",
            "Permanent address": "Adresa",
            "Contact address": "Kontaktní adresa",
            "Field is required": "Toto pole je povinné",
            "Addresses differ": "Zadat jinou kontaktní adresu",
            "Back": "Předchozí",
            "Next": "Následující",
            "Rows per page": "Počet záznamů na stránce",
            "Of": "z",
            "Delete confirmation": "Opravdu si přejete záznam smazat?",
            "Delete confirmation text": "Uživatel {{fullName}} bude smazán trvale. Akce nemůže být vrácen zpět.",
        }
    }
};

i18n
    .use(initReactI18next) // passes i18n down to react-i18next
    .init({
        resources,

        fallbackLng: ['en', 'cs'],

        lng: 'cs',

        keySeparator: false, // we do not use keys in form messages.welcome

        interpolation: {
            escapeValue: false // react already safes from xss
        }
    });

export default i18n;
