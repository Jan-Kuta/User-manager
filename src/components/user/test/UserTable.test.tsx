import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import {UserTable} from '../UserTable';
import {userList} from "./MockApplicationState";
import {initReactI18next} from "react-i18next";
import i18n from "../../../i18n";

describe("User table", () => {
    initReactI18next.init(i18n);

    it("it should match", () => {
        const renderer = ShallowRenderer.createRenderer();

        renderer.render(<UserTable users={userList}/>);
        const result = renderer.getRenderOutput();

        expect(result).toMatchSnapshot();
    });
});
