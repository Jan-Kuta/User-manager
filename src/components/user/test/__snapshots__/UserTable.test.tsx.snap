// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User table it should match 1`] = `
<React.Fragment>
  <ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)>
      <div>
        <UNDEFINED />
        <ForwardRef(WithStyles)
          component="div"
          id="tableTitle"
          variant="h6"
        >
          Seznam uživatelů
        </ForwardRef(WithStyles)>
      </div>
      <ForwardRef
        to="/users/create"
      >
        <ForwardRef(WithStyles)
          placement="left"
          title="Přidat uživatele"
        >
          <ForwardRef(WithStyles)
            aria-label="add user"
          >
            <UNDEFINED />
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
      </ForwardRef>
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        aria-label="collapsible table"
        size="small"
        stickyHeader={true}
      >
        <ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)>
            <ForwardRef(WithStyles) />
            <ForwardRef(WithStyles)
              align="right"
            >
              Jméno
            </ForwardRef(WithStyles)>
            <ForwardRef(WithStyles)
              align="right"
            >
              Příjmení
            </ForwardRef(WithStyles)>
            <Hidden
              xsDown={true}
            >
              <ForwardRef(WithStyles)
                align="right"
              >
                Uživatelské jméno
              </ForwardRef(WithStyles)>
            </Hidden>
            <ForwardRef(WithStyles)
              align="right"
            >
              Stav
            </ForwardRef(WithStyles)>
            <Hidden
              xsDown={true}
            >
              <ForwardRef(WithStyles)
                align="right"
              >
                Akce
              </ForwardRef(WithStyles)>
            </Hidden>
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)>
          <UserTableRow
            openUserDeleteDialog={[Function]}
            openUserEditDialog={[Function]}
            user={
              Object {
                "contactAddress": Object {
                  "city": "Nejdek",
                  "street": "Kollárova 964",
                  "zip": "362 21",
                },
                "email": "jan.kuta@email.cz",
                "firstName": "Jan",
                "id": "u1",
                "lastName": "Kuta",
                "permanentAddress": Object {
                  "city": "Ústí nad Labem",
                  "street": "Žežická 39",
                  "zip": "400 07",
                },
                "phone": "+420 775 246 369",
                "state": "STATE_NEW",
                "username": "Kuticzech",
              }
            }
          />
          <UserTableRow
            openUserDeleteDialog={[Function]}
            openUserEditDialog={[Function]}
            user={
              Object {
                "email": "huggo.tester@gmail.com",
                "firstName": "Huggo",
                "id": "u2",
                "lastName": "Tester",
                "permanentAddress": Object {
                  "city": "Kamenice",
                  "street": "Vymyšlená 39",
                  "zip": "100 01",
                },
                "phone": "+420 777 246 369",
                "state": "STATE_VALIDATED",
                "username": "Huggo",
              }
            }
          />
          <UserTableRow
            openUserDeleteDialog={[Function]}
            openUserEditDialog={[Function]}
            user={
              Object {
                "email": "jana.kutova@email.cz",
                "firstName": "Jana",
                "id": "u3",
                "lastName": "Kutová",
                "permanentAddress": Object {
                  "city": "Nejdek",
                  "street": "Kollárova 964",
                  "zip": "362 21",
                },
                "phone": "+420 775 246 369",
                "state": "STATE_VIP",
                "username": "Janciii",
              }
            }
          />
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      backIconButtonText="Předchozí"
      component="div"
      count={3}
      labelDisplayedRows={[Function]}
      labelRowsPerPage="Počet záznamů na stránce"
      nextIconButtonText="Následující"
      onChangePage={[Function]}
      onChangeRowsPerPage={[Function]}
      page={0}
      rowsPerPage={5}
      rowsPerPageOptions={
        Array [
          5,
          10,
          25,
        ]
      }
    />
  </ForwardRef(WithStyles)>
</React.Fragment>
`;
